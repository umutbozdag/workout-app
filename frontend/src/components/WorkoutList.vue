<template>
  <div class="workout-list-container">
    <el-row type="flex" justify="space-between" align="middle">
      <el-col>
        <h2>Workouts</h2>
      </el-col>

      <el-col :span="12" :offset="12">
        <el-button
          type="primary"
          size="small"
          plain
          icon="el-icon-plus"
          @click="pushToAddWorkout"
          >Add workout</el-button
        >
      </el-col>
    </el-row>
    <el-input
      v-model="searchText"
      prefix-icon="el-icon-search"
      placeholder="Workout name"
      clearable
    >
    </el-input>

    <el-divider></el-divider>

    <div
      v-if="computedWorkoutsData && computedWorkoutsData.length"
      class="workout-list"
    >
      <workout-list-item
        v-for="workout in computedWorkoutsData"
        :key="workout.workoutId"
        :title="workout.title"
        :description="workout.description"
        :completed="workout.exercises.every((exercise) => exercise.completed)"
        :workout-id="workout.workoutId"
        :exerciseCount="workout.exercises.length"
      />
    </div>

    <el-empty description="No workouts." v-else>
      <el-button type="primary" @click="pushToAddWorkout"
        >Add Workout</el-button
      >
    </el-empty>
  </div>
</template>

<script lang="ts">
import Vue from "vue";
import { mapActions, mapGetters, mapState } from "vuex";
import { IWorkout } from "../interfaces";
import WorkoutListItem from "./WorkoutListItem.vue";

export default Vue.extend({
  name: "WorkoutList",
  components: { WorkoutListItem },
  data() {
    return {
      searchText: "",
    };
  },
  async mounted() {
    await this.getWorkouts();
  },
  watch: {
    searchText(newVal: string) {
      this.setWorkoutSearchText(newVal);
    },
  },
  methods: {
    ...mapActions(["getWorkouts", "setWorkoutSearchText"]),
    pushToAddWorkout() {
      this.$router.push({ name: "WorkoutAdd" });
    },
  },
  computed: {
    ...mapState(["workouts"]),
    ...mapGetters(["filteredWorkouts"]),
    computedWorkoutsData() {
      if (this.filteredWorkouts && this.filteredWorkouts) {
        return this.filteredWorkouts;
      } else {
        return this.workouts;
      }
    },
  },
});
</script>

<style lang="scss" scoped>
.workout-list {
  height: 500px;
  overflow-y: auto;
  &-item {
    margin-bottom: 18px;
    margin-right: 12px;
  }
}
</style>
