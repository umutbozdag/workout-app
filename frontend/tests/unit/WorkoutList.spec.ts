import { shallowMount, createLocalVue, mount } from "@vue/test-utils";
import Vuex from "vuex";
import WorkoutList from "@/components/WorkoutList.vue";
import { generateMockData } from "./generateMockData";

const localVue = createLocalVue();
localVue.use(Vuex);

const mockWorkoutsData = generateMockData("workouts", 10);

describe("WorkoutList.vue", () => {
  let store: any;

  beforeEach(() => {
    store = new Vuex.Store({
      actions: {
        getWorkouts: jest.fn(),
      },
      getters: {
        filteredWorkouts: jest.fn(),
      },
    });
  });

  it("Renders WorkoutListItem component in it correctly if 'workouts' array has items", () => {
    const wrapper = shallowMount(WorkoutList, {
      store,
      localVue,
      computed: {
        workouts() {
          return mockWorkoutsData;
        },
      },
    });

    const workoutListComp = wrapper.findComponent({ name: "WorkoutListItem" }); // => finds WorkoutListItem by `name`
    expect(workoutListComp.exists()).toBe(true);
  });

  it("Renders empty component if 'workouts' array is empty", () => {
    const wrapper = mount(WorkoutList, {
      store,
      localVue,
      computed: {
        workouts() {
          return [];
        },
      },
    });

    const workoutListComp = wrapper.findComponent({ name: "WorkoutListItem" }); // => finds WorkoutListItem by `name`
    expect(workoutListComp.exists()).toBe(false);
  });

  it("Renders correctly with 'workouts' data", () => {
    const wrapper = mount(WorkoutList, {
      store,
      localVue,
      computed: {
        workouts() {
          return mockWorkoutsData;
        },
      },
    });

    expect(wrapper.element).toMatchSnapshot();
  });
});
