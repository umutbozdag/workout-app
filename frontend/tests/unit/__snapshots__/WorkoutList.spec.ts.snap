// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WorkoutList.vue Renders correctly with 'workouts' data 1`] = `
<div
  class="workout-list-container"
>
  <div
    class="el-row is-justify-space-between is-align-middle el-row--flex"
  >
    <div
      class="el-col el-col-24"
    >
      <h2>
        Workouts
      </h2>
    </div>
     
    <div
      class="el-col el-col-12 el-col-offset-12"
    >
      <button
        class="el-button el-button--primary el-button--small is-plain"
        type="button"
      >
        <!---->
        <i
          class="el-icon-plus"
        />
        <span>
          Add workout
        </span>
      </button>
    </div>
  </div>
   
  <div
    class="el-input el-input--prefix el-input--suffix"
  >
    <!---->
    <input
      autocomplete="off"
      class="el-input__inner"
      placeholder="Workout name"
      type="text"
    />
    <span
      class="el-input__prefix"
    >
      <i
        class="el-input__icon el-icon-search"
      />
    </span>
    <!---->
    <!---->
    <!---->
  </div>
   
  <div
    class="el-divider el-divider--horizontal"
  >
    <!---->
  </div>
   
  <div
    class="workout-list"
  >
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
    <div
      class="workout-list-item"
    >
      <div
        class="workout-list-item__info"
      >
        <h3
          class="workout-list-item__title"
        >
          Workout Title
        </h3>
         
        <p
          class="workout-list-item__desc"
        >
          Lorem ipsum dolor sit amet, consectetur adipiscing elit.
        </p>
         
        <p
          class="workout-list-item__completed"
        >
          
      Workout is completed
    
        </p>
         
        <div
          class="workout-list-item__exercise-count"
        >
          <!---->
           
        10 Exercises
       
          <!---->
        </div>
      </div>
       
      <i
        class="el-icon-right workout-list-item__icon"
      />
    </div>
  </div>
</div>
`;
